pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -39,284 +108,413 @@ func Add(mgr manager.Manager) error {
+					return reconcile.Result{}, err
+				}
 
-	//Continue even if there are pods not managed by a ReplicationController, ReplicaSet, Job, DaemonSet or StatefulSet.
-	//This is required because VirtualMachineInstance pods are not owned by a ReplicaSet or DaemonSet controller.
-	//This means that the drain operation can’t guarantee that the pods being terminated on the target node will get
-	//re-scheduled replacements placed else where in the cluster after the pods are evicted.
-	//KubeVirt has its own controllers which manage the underlying VirtualMachineInstance pods.
-	//Each controller behaves differently to a VirtualMachineInstance being evicted.
-	r.drainer.Force = true
+				statusAnnoData.status = NodeStateCordon
+				if node, err = r.setStatusAnnotation(node, statusAnnoData); err != nil {
+					return reconcile.Result{}, err
+				}
 
-	//Continue even if there are pods using emptyDir (local data that will be deleted when the node is drained).
-	//This is necessary for removing any pod that utilizes an emptyDir volume.
-	//The VirtualMachineInstance Pod does use emptryDir volumes,
-	//however the data in those volumes are ephemeral which means it is safe to delete after termination.
-	r.drainer.DeleteLocalData = true
+			}
 
-	//Ignore DaemonSet-managed pods.
-	//This is required because every node running a VirtualMachineInstance will also be running our helper DaemonSet called virt-handler.
-	//This flag indicates that it is safe to proceed with the eviction and to just ignore DaemonSets.
-	r.drainer.IgnoreAllDaemonSets = true
+		case LeaseStatusActiveOwnedByMeAboutToExpire:
+			{
+				if specAnnoData.enableMaintWindowExtensions {
 
-	//Period of time in seconds given to each pod to terminate gracefully. If negative, the default value specified in the pod will be used.
-	r.drainer.GracePeriodSeconds = -1
+					// renew the lease now.
+					lease.Spec.AcquireTime = &metav1.MicroTime{time.Now()}
 
-	// TODO - add logical value or attach from the maintancene CR
-	//The length of time to wait before giving up, zero means infinite
-	r.drainer.Timeout = time.Minute
+					if err := r.client.Update(context.TODO(), lease); err != nil {
+						r.reqLogger.Error("Failed to update the lease %v", err)
 
-	// TODO - consider pod selectors (only for VMIs + others ?)
-	//Label selector to filter pods on the node
-	//r.drainer.PodSelector = "kubevirt.io=virt-launcher"
+						statusAnnoData.status = NodeStateFailedToExtendLease
+						_, err = r.setStatusAnnotation(node, statusAnnoData)
 
-	cs, err := kubernetes.NewForConfig(config)
-	if err != nil {
-		return err
-	}
-	r.drainer.Client = cs
-	r.drainer.DryRun = false
+						return reconcile.Result{}, err
+					}
+				}
+			}
+			fallthrough
 
-	return nil
-}
+		case LeaseStatusActiveOwnedByMe:
+			{
 
-var _ reconcile.Reconciler = &ReconcileNodeMaintenance{}
+				if statusAnnoData.status != NodeStateCordon && statusAnnoData.status != NodeStateEvicting {
```

Comment:
Ok its illegal, but can we do about it?
Blocking until the lease expires doesn't seem like a great option

Reply:
i am not sure what this is referring to: had to squash all previous commits because DCO signoff wouldn't pass.